#### Análise de Dados com PYTHON.
- Acessando as LINHAS 


#### * Carregando dados com PANDAS

In [2]:
pip install pandas

In [3]:
import pandas

###### A fim de evitar ficar chamando a todo momento uma biblioteca podemos atribuir a elas, apelidos, para mais rapidez e facilidade nos códigos
###### Porém estes apelidos seguem padrões dentro da comunidade PY, exemplo disso é a pandas que tem como apelido o pd

In [4]:
import pandas as pd

In [5]:
# importando dados de um arquivo csv - lembrando que o arquivo deve estar dentro da nossa pasta de desenvolvimento

dados = pd.read_csv("base_dados.csv")                    # read = ler

In [6]:
# verificando as primeiras linhas carregadas

dados.head()                                             # head = cabeçalho

numero_pedido         cliente        motorista dt_emissao_pedido  \
0            A3         WALMART     FELIPE SILVA        2019-01-27   
1            A4  MAGAZINE LUIZA   TÚLIO SILVEIRA        2019-01-28   
2            A7  MAGAZINE LUIZA  VALDIR ESPINOSA        2019-02-04   
3           A12  MAGAZINE LUIZA     MARCOS LEROY        2019-02-18   
4           A15         WALMART      LUIZ PARDAL        2019-02-21   

  estado_destino  valor_pedido  dt_entrega motivo_devolucao    status  
0             SP         139.5  2019-02-03       S/ Devolu.  No Prazo  
1             RJ         134.4  2019-02-12       S/ Devolu.  Atrasado  
2             SP          79.5  2019-02-14   Produto Errado  No Prazo  
3             SP         326.0  2019-02-25   Produto Errado  No Prazo  
4             RJ         597.6  2019-03-07       S/ Devolu.  Atrasado

In [7]:
dados.head(10)                        # verificando as 10 primeiras linhas

numero_pedido         cliente        motorista dt_emissao_pedido  \
0            A3         WALMART     FELIPE SILVA        2019-01-27   
1            A4  MAGAZINE LUIZA   TÚLIO SILVEIRA        2019-01-28   
2            A7  MAGAZINE LUIZA  VALDIR ESPINOSA        2019-02-04   
3           A12  MAGAZINE LUIZA     MARCOS LEROY        2019-02-18   
4           A15         WALMART      LUIZ PARDAL        2019-02-21   
5           A16         WALMART       JOÃO GOMES        2019-02-23   
6           A17         WALMART       JOÃO GOMES        2019-02-24   
7           A20  MAGAZINE LUIZA  VALDIR ESPINOSA        2019-02-26   
8           A22      AMERICANAS  VALDIR ESPINOSA        2019-03-01   
9           A23      AMERICANAS     MARCOS LEROY        2019-03-02   

  estado_destino  valor_pedido  dt_entrega motivo_devolucao    status  
0             SP         139.5  2019-02-03       S/ Devolu.  No Prazo  
1             RJ         134.4  2019-02-12       S/ Devolu.  Atrasado  
2             SP          79.5  2019-02-14   Produto Errado  No Prazo  
3             SP         326.0  2019-02-25   Produto Errado  No Prazo  
4             RJ         597.6  2019-03-07       S/ Devolu.  Atrasado  
5             RJ         467.6  2019-03-06   Produto Errado  Atrasado  
6             SP         297.0  2019-03-09   Produto Errado  Atrasado  
7             BH         473.2  2019-03-07   Produto Errado  No Prazo  
8             SP         165.0  2019-03-13       Danificado  Atrasado  
9             RJ          31.0  2019-03-06       Danificado  No Prazo

In [8]:
# verificar as últimas linhas carregadas

dados.tail()                                     # tail = rabo (últimas)

numero_pedido         cliente        motorista dt_emissao_pedido  \
4277         A9675  MAGAZINE LUIZA  VALDIR ESPINOSA        2021-11-27   
4278         A9676         WALMART      LUIZ PARDAL        2021-11-28   
4279         A9680    CASA E VIDEO   TÚLIO SILVEIRA        2021-12-02   
4280         A9681  RICARDO ELETRO     MARCOS LEROY        2021-12-03   
4281         A9683      AMERICANAS     FELIPE SILVA        2021-12-06   

     estado_destino  valor_pedido  dt_entrega motivo_devolucao    status  
4277             RJ         763.2  2021-12-15   Produto Errado  Atrasado  
4278             SP         226.8  2021-12-12   Produto Errado  Atrasado  
4279             SP         821.1  2021-12-20       Danificado  Atrasado  
4280             SP          49.8  2021-12-17       S/ Devolu.  Atrasado  
4281             BH         715.5  2021-12-17       Danificado  Atrasado

In [9]:
# verificando o tamanho da base de dados, que será apresentado pelo comando shape, a quantidade de linhas e de colunas

dados.shape                           # por ser uma propriedade e não um método o shape não exige os parênteses

(4282, 9)

#### Carregando dados com Excel

###### - Recomenda-se, por uma boa prática, que se faça uma análise exploratória dos dados recebidos antes de joga-lo em nosso sistema.
###### - Mas do que se trata essa análise exploratória?
######   * Abrir o arquivo recebido, saber sobre suas abas e se não há abas ocultas, se houver deve-se abri-las;
######   * Estudar e entender os dados e as colunas do arquivo recebido;
######   * Saber o número de linhas existentes nesses arquivos, no caso do excell para chegar nessa informação -> na linha A2 CTRL + SHIFT + seta p/ baixo
       

In [10]:
dados = pd.read_excel("base_dados.xlsx")

In [11]:
# Verificando as primeiras linhas

dados.head()


numero_pedido         cliente        motorista dt_emissao_pedido  \
0            A3         WALMART     FELIPE SILVA        2019-01-27   
1            A4  MAGAZINE LUIZA   TÚLIO SILVEIRA        2019-01-28   
2            A7  MAGAZINE LUIZA  VALDIR ESPINOSA        2019-02-04   
3           A12  MAGAZINE LUIZA     MARCOS LEROY        2019-02-18   
4           A15         WALMART      LUIZ PARDAL        2019-02-21   

  estado_destino  valor_pedido dt_entrega motivo_devolucao    status  
0             SP         139.5 2019-02-03       S/ Devolu.  No Prazo  
1             RJ         134.4 2019-02-12       S/ Devolu.  Atrasado  
2             SP          79.5 2019-02-14   Produto Errado  No Prazo  
3             SP         326.0 2019-02-25   Produto Errado  No Prazo  
4             RJ         597.6 2019-03-07       S/ Devolu.  Atrasado

In [12]:
# Verificando as últimas linhas

dados.tail()


numero_pedido         cliente        motorista dt_emissao_pedido  \
4277         A9675  MAGAZINE LUIZA  VALDIR ESPINOSA        2021-11-27   
4278         A9676         WALMART      LUIZ PARDAL        2021-11-28   
4279         A9680    CASA E VIDEO   TÚLIO SILVEIRA        2021-12-02   
4280         A9681  RICARDO ELETRO     MARCOS LEROY        2021-12-03   
4281         A9683      AMERICANAS     FELIPE SILVA        2021-12-06   

     estado_destino  valor_pedido dt_entrega motivo_devolucao    status  
4277             RJ         763.2 2021-12-15   Produto Errado  Atrasado  
4278             SP         226.8 2021-12-12   Produto Errado  Atrasado  
4279             SP         821.1 2021-12-20       Danificado  Atrasado  
4280             SP          49.8 2021-12-17       S/ Devolu.  Atrasado  
4281             BH         715.5 2021-12-17       Danificado  Atrasado

In [13]:
# Verificando o tamanho da base de dados

dados.shape                  # Lembrando que,shape, por ser uma propriedade e não um método não exige os parêntese no final

(4282, 9)

#### Verificando o tipo de dados das colunas

In [14]:
dados.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4282 entries, 0 to 4281
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   numero_pedido      4282 non-null   object        
 1   cliente            4282 non-null   object        
 2   motorista          4282 non-null   object        
 3   dt_emissao_pedido  4282 non-null   datetime64[ns]
 4   estado_destino     4282 non-null   object        
 5   valor_pedido       4282 non-null   float64       
 6   dt_entrega         4282 non-null   datetime64[ns]
 7   motivo_devolucao   4282 non-null   object        
 8   status             4282 non-null   object        
dtypes: datetime64[ns](2), float64(1), object(6)
memory usage: 301.2+ KB


     - As informações acima nos trás: 
- 4282 linhas no arquivo; 
- em 9 colunas; 
- o nome das colunas; 
- non-null, que indica que não há nenhuma linha sem dado; (no-null deve ser aprofundado, pois trat-se de uma análise mais apurada)
- e o dtype, que trás a informação de 'object', que indica que é uma strings
- datetime que é uma coluna de data e hora;
- há na tabela valores float, informações não inteiras (float64)

In [15]:
# Podemos verificar o tipo de dado da variável dados

type(dados)

pandas.core.frame.DataFrame

#### Acessando Colunas individualmente
- Dados tabular -> trata-se de uma tabela com uma sequência de linhas e colunas.

In [16]:
dados.head()

numero_pedido         cliente        motorista dt_emissao_pedido  \
0            A3         WALMART     FELIPE SILVA        2019-01-27   
1            A4  MAGAZINE LUIZA   TÚLIO SILVEIRA        2019-01-28   
2            A7  MAGAZINE LUIZA  VALDIR ESPINOSA        2019-02-04   
3           A12  MAGAZINE LUIZA     MARCOS LEROY        2019-02-18   
4           A15         WALMART      LUIZ PARDAL        2019-02-21   

  estado_destino  valor_pedido dt_entrega motivo_devolucao    status  
0             SP         139.5 2019-02-03       S/ Devolu.  No Prazo  
1             RJ         134.4 2019-02-12       S/ Devolu.  Atrasado  
2             SP          79.5 2019-02-14   Produto Errado  No Prazo  
3             SP         326.0 2019-02-25   Produto Errado  No Prazo  
4             RJ         597.6 2019-03-07       S/ Devolu.  Atrasado

In [17]:
# Acessando a coluna cliente

dados["cliente"]

0              WALMART
1       MAGAZINE LUIZA
2       MAGAZINE LUIZA
3       MAGAZINE LUIZA
4              WALMART
             ...      
4277    MAGAZINE LUIZA
4278           WALMART
4279      CASA E VIDEO
4280    RICARDO ELETRO
4281        AMERICANAS
Name: cliente, Length: 4282, dtype: object

In [18]:
# Acessando a coluna número pedido

dados["numero_pedido"]

0          A3
1          A4
2          A7
3         A12
4         A15
        ...  
4277    A9675
4278    A9676
4279    A9680
4280    A9681
4281    A9683
Name: numero_pedido, Length: 4282, dtype: object

In [19]:
# Acessando multiplas colunas

dados[["cliente", "estado_destino"]]

cliente estado_destino
0            WALMART             SP
1     MAGAZINE LUIZA             RJ
2     MAGAZINE LUIZA             SP
3     MAGAZINE LUIZA             SP
4            WALMART             RJ
...              ...            ...
4277  MAGAZINE LUIZA             RJ
4278         WALMART             SP
4279    CASA E VIDEO             SP
4280  RICARDO ELETRO             SP
4281      AMERICANAS             BH

[4282 rows x 2 columns]

In [20]:
# Dentro da análise podemos criar uma lista recebendo o nome de uma varíavel e solicita-la no Data Frame.

colunas = ["cliente", "motorista", "estado_destino","status"]

dados[colunas]

cliente        motorista estado_destino    status
0            WALMART     FELIPE SILVA             SP  No Prazo
1     MAGAZINE LUIZA   TÚLIO SILVEIRA             RJ  Atrasado
2     MAGAZINE LUIZA  VALDIR ESPINOSA             SP  No Prazo
3     MAGAZINE LUIZA     MARCOS LEROY             SP  No Prazo
4            WALMART      LUIZ PARDAL             RJ  Atrasado
...              ...              ...            ...       ...
4277  MAGAZINE LUIZA  VALDIR ESPINOSA             RJ  Atrasado
4278         WALMART      LUIZ PARDAL             SP  Atrasado
4279    CASA E VIDEO   TÚLIO SILVEIRA             SP  Atrasado
4280  RICARDO ELETRO     MARCOS LEROY             SP  Atrasado
4281      AMERICANAS     FELIPE SILVA             BH  Atrasado

[4282 rows x 4 columns]

#### Acessando as Linhas

In [21]:
dados.head()

numero_pedido         cliente        motorista dt_emissao_pedido  \
0            A3         WALMART     FELIPE SILVA        2019-01-27   
1            A4  MAGAZINE LUIZA   TÚLIO SILVEIRA        2019-01-28   
2            A7  MAGAZINE LUIZA  VALDIR ESPINOSA        2019-02-04   
3           A12  MAGAZINE LUIZA     MARCOS LEROY        2019-02-18   
4           A15         WALMART      LUIZ PARDAL        2019-02-21   

  estado_destino  valor_pedido dt_entrega motivo_devolucao    status  
0             SP         139.5 2019-02-03       S/ Devolu.  No Prazo  
1             RJ         134.4 2019-02-12       S/ Devolu.  Atrasado  
2             SP          79.5 2019-02-14   Produto Errado  No Prazo  
3             SP         326.0 2019-02-25   Produto Errado  No Prazo  
4             RJ         597.6 2019-03-07       S/ Devolu.  Atrasado

###### Os números sequênciais nas linhas, são índices criados automaticamente pelo carregamento do sistema e são idependentes das sequências da página do excel.
###### PANDAS trata-se de uma biblioteca muito poderosa e entre esses poderes há várias formas de se chamar dados. Porém como estamos no aprendizado de melhores práticas, vamos fazer a seleção de linhas pelo método loc (localization).

In [22]:
dados.loc[0]

numero_pedido                         A3
cliente                          WALMART
motorista                   FELIPE SILVA
dt_emissao_pedido    2019-01-27 00:00:00
estado_destino                        SP
valor_pedido                       139.5
dt_entrega           2019-02-03 00:00:00
motivo_devolucao              S/ Devolu.
status                          No Prazo
Name: 0, dtype: object

In [23]:
dados.loc[77]

numero_pedido                       A185
cliente                       AMERICANAS
motorista                    LUIZ PARDAL
dt_emissao_pedido    2019-07-05 00:00:00
estado_destino                        BH
valor_pedido                       806.4
dt_entrega           2019-07-18 00:00:00
motivo_devolucao          Produto Errado
status                          Atrasado
Name: 77, dtype: object

In [27]:
# Selecionar linhas e colunas com loc

dados.loc[0, "cliente"]

'WALMART'

In [28]:
dados.loc[0, "motorista"]

'FELIPE SILVA'

In [32]:
dados.loc[0, ["cliente", "motorista"]]

cliente           WALMART
motorista    FELIPE SILVA
Name: 0, dtype: object